staging/src/k8s.io/apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,194 @@
+	s.removeGroupResourceFor(r)
+}
+
+func (s *DefaultManager) removeGroupResourceFor(r *ResourceInfo) {
+	gvrs := r.EquivalentResourceMapper.EquivalentResourcesFor(schema.GroupVersionResource{
+		Group:    r.Group,
+		Resource: r.Resource.Name,
+	}, "")
+	for _, gvr := range gvrs {
+		s.removeGroupResource(gvr.Group, gvr.Version)
+	}
+}
+
+func (s *DefaultManager) removeGroupResource(group, resource string) {
+	if _, ok := s.groupResources[group]; !ok {
+		return
+	}
+	delete(s.groupResources[group], resource)
+	if len(s.groupResources[group]) == 0 {
+		delete(s.groupResources, group)
+	}
+}
+
+// UpdatesPending returns if the StorageVersion of a resource is still wait to be updated.
+func (s *DefaultManager) UpdatesPending(group, resource string) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if _, ok := s.groupResources[group]; !ok {
+		return false
+	}
+	_, ok := s.groupResources[group][resource]
+	return ok
+}
+
+// DefaultManager indicates if the aggregator, kube-apiserver, and the
+// apiextensions-apiserver have completed reporting their storage versions.
+type DefaultManager struct {
+	completed atomic.Value
+
+	mu             sync.RWMutex
+	resources      map[*ResourceInfo]struct{}
+	groupResources map[string]map[string]struct{}
+}
+
+// setComplete marks the completion of updating StorageVersions. No write requests need to be blocked anymore.
+func (s *DefaultManager) setComplete() {
+	s.completed.Store(true)
+}
+
+// Completed returns if updating StorageVersions has completed.
+func (s *DefaultManager) Completed() bool {
+	return s.completed.Load().(bool)
+}
+
+func decodableVersions(e runtime.EquivalentResourceRegistry, group string, resource string) []string {
+	var versions []string
+	decodingGVRs := e.EquivalentResourcesFor(schema.GroupVersionResource{
+		Group:    group,
+		Resource: resource,
+	}, "")
+	for _, v := range decodingGVRs {
+		versions = append(versions, v.GroupVersion().String())
+	}
+	return versions
+}
+
+// UpdateStorageVersions updates the StorageVersions. If the updates are
+// successful, following calls to Completed() returns true.
+func (s *DefaultManager) UpdateStorageVersions(loopbackClientConfig *rest.Config, serverID string) {
+	cfg := rest.AddUserAgent(loopbackClientConfig, "system:kube-apiserver")
+	clientset, err := apiserverclientset.NewForConfig(cfg)
+	if err != nil {
+		klog.Fatalf("failed to get clientset: %v", err)
+		return
+	}
+	sc := clientset.InternalV1alpha1().StorageVersions()
+
+	s.mu.RLock()
+	resources := s.resources
+	s.mu.RUnlock()
+	for r := range resources {
+		r.DecodableVersions = decodableVersions(r.EquivalentResourceMapper, r.Group, r.Resource.Name)
+		if err := updateStorageVersionFor(sc, serverID, r.Group+"."+r.Resource.Name, r.EncodingVersion, r.DecodableVersions); err != nil {
+			klog.Fatalf("failed to update storage version for %v", r.Resource.Name)
+			return
+		}
+		klog.V(2).Infof("successfully updated storage version for %v", r.Resource.Name)
+		s.RemoveResourceInfo(r)
```

Comment:
I suspect we will want to know the full list of ones we handled.  Instead, can you mark it as completed.